x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/analytics/EnumCounters.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.analytics;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLongArray;
+
+/**
+ * Utility class similar to org.elasticsearch.xpack.core.watcher.common.stats.Counters, but it is using Enum instead
+ * of string to identify the counter.
+ */
+public class EnumCounters<E extends Enum<E>> implements Writeable {
+    private final AtomicLongArray counters;
+    private final E[] enums;
+
+    public EnumCounters(Class<E> enumClass) {
+        counters = new AtomicLongArray(enumClass.getEnumConstants().length);
+        enums = enumClass.getEnumConstants();
+    }
+
+    public EnumCounters(StreamInput in, Class<E> enumClass) throws IOException {
+        int size = in.readVInt();
+        enums = enumClass.getEnumConstants();
+        long[] vals = new long[enums.length];
+        for (int i = 0; i < size; i++) {
+            long val = in.readVLong();
+            if (i < vals.length) {
+                vals[i] = val;
+            }
+        }
+        counters = new AtomicLongArray(vals);
```

Comment:
So, like, we have to add enum entries to the end and they'll come through as 0 if they aren't set. Right?

Reply:
I think it is worth adding comments about that, both to the enum class and to this one.

Reply:
This is basically to handle the situation when an older node have less stat categories or a newer node has more stats categories. I added a comment to the [enum](https://github.com/elastic/elasticsearch/pull/54911/files/c6818b96a9c7c4524079bc63835b3c1284eec523#diff-e6deefb95c3fa58ba7be8b2b36cb766dR36) but I agree I should duplicate it here.

Reply:
:+1: